### PR TITLE
Initial throttling support

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

Handles retries with exponential backoff for WhyLabs REST endpoints and s3 uploads.

## Changes

- Wrap whylabs-client requests in `@backoff` to retry on failure
- Trying out HTTPretty REST fakes for testing

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

**Review by catel22 — DISMISSED** [review]
Looks good!

File: python/tests/api/writer/test_whylabs_integration.py [hunk 1]
```diff
@@ -39,6 +41,40 @@
...
+        writer.write(result.profile())
+    except ApiException as e:
+        assert e.status == 429
+    except Exception as e:
+        assert False
+    else:
+        assert False
+"""
+
+
```
Comment (richard-rogers) [inline python/tests/api/writer/test_whylabs_integration.py]:
```suggestion
```

**Review by w0-automator — DISMISSED** [review]

**Review by jamie256 — DISMISSED** [review]
LGTM!

File: python/whylogs/api/writer/whylabs_client.py [hunk 2]
```diff
@@ -425,8 +432,20 @@ def tag_custom_performance_column(
             column=column,
             default_metric=default_metric,
         )
+
+        @backoff.on_exception(
+            backoff.expo,
+            ApiException,
+            giveup=lambda e: e.status != 429,  # type: ignore
```
Comment (jamie256) [inline python/whylogs/api/writer/whylabs_client.py:439]: Do we not retry on 500?

Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs_client.py:439]: Not at present

Comment (jamie256) [reply python/whylogs/api/writer/whylabs_client.py:439]: We should also retry on 5xx, but we can do that in a follow up.

**Review by w0-automator — APPROVED** [review]
LGTM!

**Pull request merged**